### PR TITLE
feat(protocol-designer): add heater shaker step form

### LINKED ISSUE #9689: H-S Step Form
# Overview

This ticket is to build out the H-S Step Form. The labware specified for the module in the protocol is the only available selection.  

# Implementation Details

**Step form contains:**

- Module dropdown with labware selection

**Set temperature toggle.** 
- Uses toggle icon
- "Deactivated' when not selected. No input field displayed if deactivated
- "Active" when selected. 
- If "active," show temp input field in Celcius, accepting temp value up to one decimal
- Acceptable range is 37-95 C
- Blurring & Validation: temp field functions just like thermocycler temp field
- Masking & Validation: repeat strategy from thermocycler temp field 

**Set shake speed toggle**
- Uses toggle icon
- "Deactivated' when not selected. No input field displayed if deactivated
- "Active" when selected. 
- If "active," show shake input field in RPM, accepting whole numbers only
- Acceptable range is 200-3000 RPM
- Blurring & Validation: same strategy as temp field
- Masking & Validation: no decimals, whole numbers only
- If shake speed is active, labware latch will be set to "disabled" and "Closed and Locked"

**Labware Latch**
- Uses toggle icon
- "Closed" is default
- "Open" when activated
- Disabled in shake speed is active, with status of "Closed and Locked" 

**Set Timer**
- Uses checkbox
- If unselected, no timer input fields display
- If selected, display two input fields: "m" (minutes field), "s" (seconds field)
- Validation: If step is saved and timer field has not been filled out, that time field is set to 0 
- Masking: Functions like thermocycler time field

When a new H-S step is created:
- The set timer form is UNCHECKED
- Temp field is deactivated
- Shake field is deactivated
- Labware Latch is set to CLOSED

Standard step form items: 
- Contains CTAs (Delete, Notes, Cancel, Save)
- Adheres to design and layout linked int his ticket

# Design Links
- https://www.figma.com/file/695ogls23YsQYImRcGYDE1/Heater-Shaker?node-id=2770%3A83961

# Acceptance Criteria

Step form contains:
- [ ] Module dropdown with labware selection

**Set temperature toggle.** 
- [ ] Uses toggle icon
- [ ] "Deactivated' when not selected. No input field displayed if deactivated
- [ ] "Active" when selected. 
- [ ] If "active," show temp input field in Celcius, accepting temp value up to one decimal
- [ ] Acceptable range is 37-95 C
- [ ] Blurring & Validation: temp field functions just like thermocycler temp field
- [ ] Masking & Validation: repeat strategy from thermocycler temp field 

**Set shake speed toggle**
- [ ] Uses toggle icon
- [ ] "Deactivated' when not selected. No input field displayed if deactivated
- [ ] "Active" when selected. 
- [ ] If "active," show shake input field in RPM, accepting whole numbers only
- [ ] Acceptable range is 200-3000 RPM
- [ ] Blurring & Validation: same strategy as temp field
- [ ] Masking & Validation: no decimals, whole numbers only
- [ ] If shake speed is active, labware latch will be set to "disabled" and "Closed and Locked"

**Labware Latch**
- [ ] Uses toggle icon
- [ ] "Closed" is default
- [ ] "Open" when activated
- [ ] Disabled in shake speed is active, with status of "Closed and Locked" 

**Set Timer**
- [ ] Uses checkbox
- [ ] If unselected, no timer input fields display
- [ ] If selected, display two input fields: "m" (minutes field), "s" (seconds field)
- [ ] Validation: If step is saved and timer field has not been filled out, that time field is set to 0 
- [ ] Masking: Functions like thermocycler time field

**When a new H-S step is created:**
- [ ] The set timer form is UNCHECKED
- [ ] Temp field is deactivated
- [ ] Shake field is deactivated
- [ ] Labware Latch is set to CLOSED

**Standard step form items:** 
- [ ] Contains CTAs (Delete, Notes, Cancel, Save)
- [ ] Adheres to design and layout linked int his ticket

### PR DESCRIPTION
**Pull request by shlokamin**
# Overview

This PR adds the heater shaker step form to PD. 

closes #9689

CO-authored with @sakibh @jerader 

# Changelog

- Adds heater shaker step form

# Review requests

- Create new heater shaker step
- Step form matches figma design

# Risk assessment

low

### CODE REVIEW

File: protocol-designer/src/components/StepEditForm/forms/HeaterShakerForm/StateFields.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+interface Props {
+  propsForFields: FieldPropsByName
+  isEndingHold?: boolean
+  formData: FormData
+}
+
+export const StateFields = (props: Props): JSX.Element => {
+  const { isEndingHold, propsForFields, formData } = props
+
+  // Append 'Hold' to field names if component is used for an ending hold in a TC profile
```
Comment (shlokamin) [inline protocol-designer/src/components/StepEditForm/forms/HeaterShakerForm/StateFields.tsx]: think this comment got copied and is leftover

File: protocol-designer/src/steplist/formLevel/handleFormChange/dependentFieldsUpdateHeaterShaker.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+// TODO: Ian 2019-02-21 import this from a more central place - see #2926
+const getDefaultFields = (...fields: StepFieldName[]): FormPatch =>
+  pick(getDefaultsForStepType('heaterShaker'), fields)
+
+const updatePatchOnHeaterShakerFormType = (
+  patch: FormPatch,
+  rawForm: FormData
+): FormPatch => {
+  // Profile => State
```
Comment (shlokamin) [inline protocol-designer/src/steplist/formLevel/handleFormChange/dependentFieldsUpdateHeaterShaker.ts]: think this comment is leftover from thee TC patcher

File: protocol-designer/src/steplist/formLevel/handleFormChange/dependentFieldsUpdateHeaterShaker.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  originalPatch: FormPatch,
+  rawForm: FormData // raw = NOT hydrated
+): FormPatch {
+  // sequentially modify parts of the patch until it's fully updated
+  return chainPatchUpdaters(originalPatch, [
+    chainPatch => updatePatchOnHeaterShakerFormType(chainPatch, rawForm),
+    chainPatch => updatePatchOnBlockChange(chainPatch, rawForm),
+    chainPatch => updatePatchOnLidChange(chainPatch, rawForm),
+  ])
+}
```
Comment (shlokamin) [inline protocol-designer/src/steplist/formLevel/handleFormChange/dependentFieldsUpdateHeaterShaker.ts]: i dont think we want any of this, or at least not yet. also most of this only applies to the TC

**Review by jerader — APPROVED** [review]
🌵

**Review by sakibh — APPROVED** [review]
✅

**Pull request merged**